### ISSUE TITLE
Incorrect uses of @inbounds cause miscalculation of statistics

### ISSUE THREAD

**Issue by yurivish** [JuliaStats/Distributions.jl#1265]
There are many incorrect uses of `@inbounds` in this package, leading to incorrect results when fitting distributions and doing other calculations on arrays with offset axes. 

I came across this issue [previously](https://github.com/JuliaStats/Distributions.jl/issues/1253) with `DiscreteUniform` but did not realize how widespread the problem was until I ran a [search](https://github.com/JuliaStats/Distributions.jl/search?q=inbounds) for `@inbounds` across the code in this package. 

Many uses of `@inbounds` in this package are in functions that accept arbitrary abstract vectors, matrices, or arrays, then disable bounds checks and index into the array with unsafe one-based indices.

One example is fitting a Normal distribution:

```julia
julia> using Distributions, OffsetArrays

julia> a = collect(-5:5);

julia> b = OffsetArray(a, -5:5);

julia> suffstats(Normal, a)
Distributions.NormalStats(0.0, 0.0, 110.0, 11.0)

julia> fit(Normal, a)
Normal{Float64}(μ=0.0, σ=3.1622776601683795)

julia> suffstats(Normal, b)
Distributions.NormalStats(7.39673056934e11, 6.7243005175818184e10, 1.5901446116594743e23, 11.0)

julia> fit(Normal, b)
Normal{Float64}(μ=6.7243005175818184e10, σ=1.2023252515852448e11)
```

https://github.com/JuliaStats/Distributions.jl/blob/863844c88e4153af13996f571fcc612d159de542/src/univariate/continuous/normal.jl#L254-L271

```
julia> versioninfo()
Julia Version 1.5.3
Commit 788b2c77c1 (2020-11-09 13:37 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin18.7.0)
  CPU: Intel(R) Core(TM) i9-9980HK CPU @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
```

**Comment by ParadaCarleton** [comment]
Are there any uses of inbounds left after [this PR](https://github.com/JuliaStats/Distributions.jl/pull/1742), @itsdebartha?

**Comment by itsdebartha** [comment]
@ParadaCarleton Yes, there are still some uses of inbounds in `categorical`, `discretenonparametric` and `poissonbinomial` for the `univariates`. They are a bit tricky to tackle...

**Comment by andreasnoack** [comment]
@itsdebartha Maybe we should just add `require_one_based_indexing` to the remaining definitions for the time being similarly to what was done in https://github.com/JuliaLang/LinearAlgebra.jl/blob/c33045c7625ad46679b6588a688f62c9a8d27aba/src/generic.jl#L782 when non-one-based indexing was introduced. Given that you've already looked into this before, would you be able to prepare a PR?
